<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.5.2" />
<title>TransDecoder (Find Coding Regions Within Transcripts)</title>
<style type="text/css">
/* Debug borders */
p, li, dt, dd, div, pre, h1, h2, h3, h4, h5, h6 {
/*
  border: 1px solid red;
*/
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

tt {
  color: navy;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  font-family: sans-serif;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}

div.sectionbody {
  font-family: serif;
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}

pre {
  padding: 0;
  margin: 0;
}

span#author {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1.1em;
}
span#email {
}
span#revnumber, span#revdate, span#revremark {
  font-family: sans-serif;
}

div#footer {
  font-family: sans-serif;
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
div#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
div#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

div#preamble {
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.tableblock, div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.0em;
  margin-bottom: 2.0em;
  margin-right: 10%;
  color: #606060;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}

div.listingblock > div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock, div.verseblock {
  padding-left: 1.0em;
  margin-left: 1.0em;
  margin-right: 10%;
  border-left: 5px solid #dddddd;
  color: #777777;
}

div.quoteblock > div.attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock > div.content {
  white-space: pre;
}
div.verseblock > div.attribution {
  padding-top: 0.75em;
  text-align: left;
}
/* DEPRECATED: Pre version 8.2.7 verse style literal block. */
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 3px solid #dddddd;
}

div.exampleblock > div.content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

div.imageblock div.content { padding-left: 0; }
span.image img { border-style: none; }
a.image:visited { color: white; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
  color: navy;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
ol.arabic {
  list-style-type: decimal;
}
ol.loweralpha {
  list-style-type: lower-alpha;
}
ol.upperalpha {
  list-style-type: upper-alpha;
}
ol.lowerroman {
  list-style-type: lower-roman;
}
ol.upperroman {
  list-style-type: upper-roman;
}

div.compact ul, div.compact ol,
div.compact p, div.compact p,
div.compact div, div.compact div {
  margin-top: 0.1em;
  margin-bottom: 0.1em;
}

div.tableblock > table {
  border: 3px solid #527bbd;
}
thead, p.table.header {
  font-family: sans-serif;
  font-weight: bold;
}
tfoot {
  font-weight: bold;
}
td > div.verse {
  white-space: pre;
}
p.table {
  margin-top: 0;
}
/* Because the table frame attribute is overriden by CSS in most browsers. */
div.tableblock > table[frame="void"] {
  border-style: none;
}
div.tableblock > table[frame="hsides"] {
  border-left-style: none;
  border-right-style: none;
}
div.tableblock > table[frame="vsides"] {
  border-top-style: none;
  border-bottom-style: none;
}


div.hdlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hdlist tr {
  padding-bottom: 15px;
}
dt.hdlist1.strong, td.hdlist1.strong {
  font-weight: bold;
}
td.hdlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
  color: navy;
}
td.hdlist2 {
  vertical-align: top;
}
div.hdlist.compact tr {
  margin: 0;
  padding-bottom: 0;
}

.comment {
  background: yellow;
}

.footnote, .footnoteref {
  font-size: 0.8em;
}

span.footnote, span.footnoteref {
  vertical-align: super;
}

#footnotes {
  margin: 20px 0 20px 0;
  padding: 7px 0 0 0;
}

#footnotes div.footnote {
  margin: 0 0 5px 0;
}

#footnotes hr {
  border: none;
  border-top: 1px solid silver;
  height: 1px;
  text-align: left;
  margin-left: 0;
  width: 20%;
  min-width: 100px;
}


@media print {
  div#footer-badges { display: none; }
}

div#toc {
  margin-bottom: 2.5em;
}

div#toctitle {
  color: #527bbd;
  font-family: sans-serif;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}
/* Workarounds for IE6's broken and incomplete CSS2. */

div.sidebar-content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}
div.sidebar-title, div.image-title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  margin-top: 0.0em;
  margin-bottom: 0.5em;
}

div.listingblock div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock-attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock-content {
  white-space: pre;
}
div.verseblock-attribution {
  padding-top: 0.75em;
  text-align: left;
}

div.exampleblock-content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

/* IE6 sets dynamically generated links as visited. */
div#toc a:visited { color: blue; }
</style>
<script type="text/javascript">
/*<![CDATA[*/
window.onload = function(){asciidoc.footnotes();}
var asciidoc = {  // Namespace.

/////////////////////////////////////////////////////////////////////
// Table Of Contents generator
/////////////////////////////////////////////////////////////////////

/* Author: Mihai Bazon, September 2002
 * http://students.infoiasi.ro/~mishoo
 *
 * Table Of Content generator
 * Version: 0.4
 *
 * Feel free to use this script under the terms of the GNU General Public
 * License, as long as you do not remove or alter this notice.
 */

 /* modified by Troy D. Hanson, September 2006. License: GPL */
 /* modified by Stuart Rackham, 2006, 2009. License: GPL */

// toclevels = 1..4.
toc: function (toclevels) {

  function getText(el) {
    var text = "";
    for (var i = el.firstChild; i != null; i = i.nextSibling) {
      if (i.nodeType == 3 /* Node.TEXT_NODE */) // IE doesn't speak constants.
        text += i.data;
      else if (i.firstChild != null)
        text += getText(i);
    }
    return text;
  }

  function TocEntry(el, text, toclevel) {
    this.element = el;
    this.text = text;
    this.toclevel = toclevel;
  }

  function tocEntries(el, toclevels) {
    var result = new Array;
    var re = new RegExp('[hH]([2-'+(toclevels+1)+'])');
    // Function that scans the DOM tree for header elements (the DOM2
    // nodeIterator API would be a better technique but not supported by all
    // browsers).
    var iterate = function (el) {
      for (var i = el.firstChild; i != null; i = i.nextSibling) {
        if (i.nodeType == 1 /* Node.ELEMENT_NODE */) {
          var mo = re.exec(i.tagName);
          if (mo && (i.getAttribute("class") || i.getAttribute("className")) != "float") {
            result[result.length] = new TocEntry(i, getText(i), mo[1]-1);
          }
          iterate(i);
        }
      }
    }
    iterate(el);
    return result;
  }

  var toc = document.getElementById("toc");
  var entries = tocEntries(document.getElementById("content"), toclevels);
  for (var i = 0; i < entries.length; ++i) {
    var entry = entries[i];
    if (entry.element.id == "")
      entry.element.id = "_toc_" + i;
    var a = document.createElement("a");
    a.href = "#" + entry.element.id;
    a.appendChild(document.createTextNode(entry.text));
    var div = document.createElement("div");
    div.appendChild(a);
    div.className = "toclevel" + entry.toclevel;
    toc.appendChild(div);
  }
  if (entries.length == 0)
    toc.parentNode.removeChild(toc);
},


/////////////////////////////////////////////////////////////////////
// Footnotes generator
/////////////////////////////////////////////////////////////////////

/* Based on footnote generation code from:
 * http://www.brandspankingnew.net/archive/2005/07/format_footnote.html
 */

footnotes: function () {
  var cont = document.getElementById("content");
  var noteholder = document.getElementById("footnotes");
  var spans = cont.getElementsByTagName("span");
  var refs = {};
  var n = 0;
  for (i=0; i<spans.length; i++) {
    if (spans[i].className == "footnote") {
      n++;
      // Use [\s\S] in place of . so multi-line matches work.
      // Because JavaScript has no s (dotall) regex flag.
      note = spans[i].innerHTML.match(/\s*\[([\s\S]*)]\s*/)[1];
      noteholder.innerHTML +=
        "<div class='footnote' id='_footnote_" + n + "'>" +
        "<a href='#_footnoteref_" + n + "' title='Return to text'>" +
        n + "</a>. " + note + "</div>";
      spans[i].innerHTML =
        "[<a id='_footnoteref_" + n + "' href='#_footnote_" + n +
        "' title='View footnote' class='footnote'>" + n + "</a>]";
      var id =spans[i].getAttribute("id");
      if (id != null) refs["#"+id] = n;
    }
  }
  if (n == 0)
    noteholder.parentNode.removeChild(noteholder);
  else {
    // Process footnoterefs.
    for (i=0; i<spans.length; i++) {
      if (spans[i].className == "footnoteref") {
        var href = spans[i].getElementsByTagName("a")[0].getAttribute("href");
        href = href.match(/#.*/)[0];  // Because IE return full URL.
        n = refs[href];
        spans[i].innerHTML =
          "[<a href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
      }
    }
  }
}

}
/*]]>*/
</script>
</head>
<body>
<div id="header">
<h1>TransDecoder (Find Coding Regions Within Transcripts)</h1>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph"><p>TransDecoder identifies candidate coding regions within transcript sequences, such as those generated by de novo RNA-Seq transcript assembly using Trinity, or constructed based on RNA-Seq alignments to the genome using Tophat and Cufflinks.</p></div>
<div class="paragraph"><p>TransDecoder identifies likely coding sequences based on the following criteria:</p></div>
<div class="ulist"><ul>
<li>
<p>
a minimum length open reading frame (ORF) is found in a transcript sequence
</p>
</li>
<li>
<p>
a log-likelihood score similar to what is computed by the GeneID software is &gt; 0.
</p>
</li>
<li>
<p>
the above coding score is greatest when the ORF is scored in the 1st reading frame as compared to scores in the other 5 reading frames.
</p>
</li>
<li>
<p>
if a candidate ORF is found fully encapsulated by the coordinates of another candidate ORF, the longer one is reported.  However, a single transcript can report multiple ORFs (allowing for operons, chimeras, etc).
</p>
</li>
<li>
<p>
<strong>optional</strong> the putative peptide has a match to a Pfam domain above the noise cutoff score.
</p>
</li>
</ul></div>
<div class="paragraph"><p>The software is primarily maintained by <a href="http://www.broadinstitute.org/~bhaas/">Brian Haas</a> at the <a href="http://broadinstitute.org">Broad Institute</a> and <a href="http://tiny.cc/alexie_pap_csiro/">Alexie Papanicolaou</a> at the <a href="http://www.csiro.au/">Commonwealth Scientific and Industrial Research Organisation</a> (CSIRO). It is integrated into other related software such as <a href="http://trinityrnaseq.sf.net">Trinity</a>, <a href="http://pasa.sf.net">PASA</a>, <a href="http://evidencemodeler.sf.net">EVidenceModeler</a>, and <a href="http://trinotate.sf.net">Trinotate</a>.</p></div>
</div>
</div>
<h2 id="_obtaining_transdecoder">Obtaining TransDecoder</h2>
<div class="sectionbody">
<div class="paragraph"><p>The latest release of TransDecoder can be found <a href="http://sourceforge.net/projects/transdecoder/">here</a>.</p></div>
</div>
<h2 id="_running_transdecoder">Running TransDecoder</h2>
<div class="sectionbody">
<h3 id="_predicting_coding_regions_from_a_transcript_fasta_file">Predicting coding regions from a transcript fasta file</h3><div style="clear:left"></div>
<div class="paragraph"><p>The <em>TransDecoder</em> utility is run on a fasta file containing the target transcript sequences.  The simplest usage is as follows:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>TransDecoder -t target_transcripts<span style="color: #990000">.</span>fasta</tt></pre></div></div>
<div class="paragraph"><p>If the transcripts are oriented according to the sense strand, then include the -S flag to examine only the top strand.  Full usage info is below.</p></div>
<div class="paragraph"><p>The script generates several output files, which are described below, but the final set of candidate coding regions can be found as files <em><strong>.transdecoder.</strong></em> where extensions include .pep, .cds, .gff3, and .bed</p></div>
<h3 id="_starting_from_a_genome_based_transcript_structure_gtf_file_eg_cufflinks">Starting from a genome-based transcript structure GTF file (eg. cufflinks)</h3><div style="clear:left"></div>
<div class="paragraph"><p>The process here is identical to the above with the exception that we must first generate a fasta file corresponding to the transcript sequences, and in the end, we recompute a genome annotation file in GFF3 format that describes the predicted coding regions in the context of the genome.</p></div>
<div class="paragraph"><p>First, convert the transcript structure GTF file to an alignment-GFF3 formatted file (this is done only because our processes operate on gff3 rather than the starting gtf file - nothing of great consequence).  Convert gtf to alignment-gff3 like so, using cufflinks GTF output as an example:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>util/cufflinks_gtf_to_alignment_gff3<span style="color: #990000">.</span>pl transcripts<span style="color: #990000">.</span>gtf <span style="color: #990000">&gt;</span> transcripts<span style="color: #990000">.</span>gff3</tt></pre></div></div>
<div class="paragraph"><p>Next, construct the transcript fasta file using the genome and the transcripts.gff3 file like so:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>util/cufflinks_gtf_genome_to_cdna_fasta<span style="color: #990000">.</span>pl transcripts<span style="color: #990000">.</span>gtf <span style="font-weight: bold"><span style="color: #0000FF">test</span></span><span style="color: #990000">.</span>genome<span style="color: #990000">.</span>fasta <span style="color: #990000">&gt;</span> transcripts<span style="color: #990000">.</span>fasta</tt></pre></div></div>
<div class="paragraph"><p>Now, run the process described above to generate your best candidate ORF predictions:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>TransDecoder -t transcripts<span style="color: #990000">.</span>fasta</tt></pre></div></div>
<div class="paragraph"><p>And finally, generate a genome-based coding region annotation file:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>cdna_alignment_orf_to_genome_orf<span style="color: #990000">.</span>pl transcripts<span style="color: #990000">.</span>fasta<span style="color: #990000">.</span>transdecoder<span style="color: #990000">.</span>gff3 transcripts<span style="color: #990000">.</span>gff3 transcripts<span style="color: #990000">.</span>fasta <span style="color: #990000">&gt;</span> transcripts<span style="color: #990000">.</span>fasta<span style="color: #990000">.</span>transdecoder<span style="color: #990000">.</span>genome<span style="color: #990000">.</span>gff3</tt></pre></div></div>
</div>
<h2 id="_sample_data_and_execution">Sample data and execution</h2>
<div class="sectionbody">
<div class="paragraph"><p>The sample_data/ directory includes a <em>runMe.sh</em> script that you can execute to demonstrate the entire process, starting from a cufflinks GTF file. Note, the typical use-case for TransDecoder is starting from a fasta file containing target <em>Transcripts</em>, however, in the case of genome analysis, transcripts are often inferred from annotation coordinates, such as in this Cufflinks GTF formatted file.  In this example, transcript sequences are reconstructed based on the GTF annotation coordinates, and then TransDecoder is executed on that fasta file.  We include an additional utility for converting the transcript ORF coordinates into genome-coordinates so these regions can be examined in the genomic context.</p></div>
</div>
<h2 id="_output_files_explained">Output files explained</h2>
<div class="sectionbody">
<div class="paragraph"><p>A temporary directory (ex. transdecoder.tmp.nopfam) is created to run and store intermediate parts of the pipeline, and contains:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>longest_orfs.pep   : all ORFs meeting the minimum length criteria, regardless of coding potential.
longest_orfs.gff3  : positions of all ORFs as found in the target transcripts
longest_orfs.cds   : the nucleotide coding sequence for all detected ORFs</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>longest_orfs.cds.top_500_longest   : the top 500 longest ORFs, used for training a Markov model for coding sequences.</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>hexamer.scores                     : log likelihood score for each k-mer  (coding/random)</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>longest_orfs.cds.scores            : the log likelihood sum scores for each ORF across each of the 6 reading frames
longest_orfs.cds.scores.selected   : the accessions of the ORFs that were selected based on the scoring criteria (described at top)
longest_orfs.cds.best_candidates.gff3               : the positions of the selected ORFs in transcripts</tt></pre>
</div></div>
<div class="paragraph"><p>Then, the final outputs are reported in your current working directory:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>transcripts.fasta.transdecoder.pep : peptide sequences for the final candidate ORFs; all shorter candidates within longer ORFs were removed.
transcripts.fasta.transdecoder.cds  : nucleotide sequences for coding regions of the final candidate ORFs
transcripts.fasta.transdecoder.gff3 : positions within the target transcripts of the final selected ORFs
transcripts.fasta.transdecoder.bed  : bed-formatted file describing ORF positions, best for viewing using GenomeView or IGV.</tt></pre>
</div></div>
</div>
<h2 id="_including_pfam_domain_searches_as_orf_retention_criteria">Including PFAM domain searches as ORF retention criteria</h2>
<div class="sectionbody">
<div class="paragraph"><p>To further maximize sensitivity for capturing ORFs that may have functional significance, regardless of coding likelihood score as mentioned above, you can scan all ORFs for PFAM domain hits, and retain all such ORFs as part of the final TransDecoder output.  Run TransDecoder like so:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>TransDecoder -t target_transcripts<span style="color: #990000">.</span>fasta --search_pfam /path/to/Pfam-A<span style="color: #990000">.</span>hmm</tt></pre></div></div>
<div class="paragraph"><p>This requires that you have <a href="http://hmmer.janelia.org/software">HMMER</a> installed and available in your PATH setting, and <a href="ftp://ftp.sanger.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz">Pfam</a> downloaded to be searched locally. If you are ok with not using the latest Pfam-A release from the Sanger institute, simply do:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>make prep_pfam</tt></pre></div></div>
<div class="paragraph"><p>This will download a <em>good-to-go</em> binary (compressed) version of the combined Pfam-A and -B datasets from the <a href="http://sourceforge.net/projects/transdecoder/files/Pfam-AB.hmm.bin/download">TransDecoder</a> site. Using Pfam-B will increase sensitivity for <em>non-model</em> species at the expense of the searches taking a bit longer. The PFAM file on our website was created using the Pfam version that was current on the Modified Date shown on <a href="http://sourceforge.net/projects/transdecoder/files/?source=navbar">Transdecoder file listing</a>.</p></div>
<div class="paragraph"><p>Alternatively, you can create a new one from scratch:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>wget ftp<span style="color: #990000">:</span>//ftp<span style="color: #990000">.</span>sanger<span style="color: #990000">.</span>ac<span style="color: #990000">.</span>uk/pub/databases/Pfam/current_release/Pfam-A<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>gz ftp<span style="color: #990000">:</span>//ftp<span style="color: #990000">.</span>sanger<span style="color: #990000">.</span>ac<span style="color: #990000">.</span>uk/pub/databases/Pfam/current_release/Pfam-B<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>gz
gunzip -c Pfam-A<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>gz Pfam-B<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>gz <span style="color: #990000">&gt;</span> Pfam-AB<span style="color: #990000">.</span>hmm
hmmconvert -b Pfam-AB<span style="color: #990000">.</span>hmm <span style="color: #990000">&gt;</span> pfam/Pfam-AB<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>bin
rm Pfam-A<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>gz Pfam-B<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>gz Pfam-AB<span style="color: #990000">.</span>hmm pfam/Pfam-AB<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>bin<span style="color: #990000">.*</span>
make prep_pfam</tt></pre></div></div>
<div class="paragraph"><p>Once either of this step is complete, you have the option of running on another single server (with ParaFly) or submit to a batch system:</p></div>
<h3 id="_stand_alone_just_use_this_for_eg_5_cpus">Stand-alone just use this for eg. 5 CPUs</h3><div style="clear:left"></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>TransDecoder -t target_transcripts<span style="color: #990000">.</span>fasta --reuse --workdir transdecoder --search_pfam /path_to_transdecoder/pfam/Pfam-AB<span style="color: #990000">.</span>hmm<span style="color: #990000">.</span>bin --CPU <span style="color: #993399">5</span></tt></pre></div></div>
<h3 id="_using_mpi">Using MPI</h3><div style="clear:left"></div>
<div class="paragraph"><p>Highly recommended: if you have a computing cluster and MPI installed, you can run TransDecoder using multiple threads with:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>TransDecoder -t target_transcripts<span style="color: #990000">.</span>fasta --search_pfam /path/to/Pfam-A<span style="color: #990000">.</span>hmm --MPI --CPU number_of_CPUs</tt></pre></div></div>
<h4 id="_how_to_prepare_for_mpi">How to prepare for MPI</h4>
<div class="ulist"><ul>
<li>
<p>
MPI is free. First you need the relevant software. We recommend openMPI. You can install it from repositories, e.g. on Debian/Ubuntu:
</p>
</li>
</ul></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>apt-get install openmpi-bin</tt></pre></div></div>
<div class="ulist"><ul>
<li>
<p>
MPICH2 specific instruction
</p>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<img src="./images/icons/tip.png" alt="Tip" />
</td>
<td class="content">If you don&#8217;t use OpenMPI but use mpich2 then use mpiexc start the MPI daemon. You only need to do this once (i.e. if you have done it before and the computer was not restarted, it should be there)</td>
</tr></table>
</div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>ps -Af<span style="color: #990000">|</span>grep <span style="color: #009900">$USER</span><span style="color: #990000">|</span>grep mpd <span style="font-style: italic"><span style="color: #9A1900"># if nothing comes back then:</span></span>
mpd <span style="color: #990000">&amp;</span></tt></pre></div></div>
</li>
</ul></div>
<div class="paragraph"><p>Then you need to re-run make so that FFINDEX is compiled correctly. Finally you need to tell the system where the FFINDEX libraries reside.
You do this by setting the environmental variable of LD_LIBRARY_PATH to point to your /full_path_to_transdecoder/util/lib64 (obviously replace the full_path_to_transdecoder):</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>make
<span style="font-weight: bold"><span style="color: #0000FF">export</span></span> <span style="color: #009900">LD_LIBRARY_PATH</span><span style="color: #990000">=</span><span style="color: #009900">$LD_LIBRARY_PATH</span><span style="color: #990000">:</span>/full_path_to_transdecoder/util/lib<span style="color: #993399">64</span></tt></pre></div></div>
<div class="paragraph"><p>You can choose to add this last command to your $HOME/.bashrc file (so it is executed automatically at login), run it manually before using --MPI with Transdecoder or ask your system administrator to do it for you.</p></div>
<h4 id="_batch_system_on_a_hpc_cluster">Batch system on a HPC cluster</h4>
<div class="paragraph"><p>For high-throughput projects, we recommend the use of a HPC cluster and the use of MPI. We regularly use it with 80-90 nodes on a PBS/Torque batch environment. The example scripts in util/pfam_mpi* are suitable for PBS and with the help of your system administrators, it is straightforward to reconfigure them for other systems (such as the popular LSF/SGI). For this scenario, ask TransDecoder to use a specific working directory, to re-use existing files and stop when the PFAM searches are prepared:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>TransDecoder -t target_transcripts<span style="color: #990000">.</span>fasta --search_pfam /path/to/Pfam-A<span style="color: #990000">.</span>hmm --MPI --workdir mytransdecoder_output --reuse --prepare_pfam
cd mytransdecoder_output
cp <span style="color: #009900">$TRANSDECODER_DIR</span>/util/pfam_mpi<span style="color: #990000">.*</span> <span style="color: #990000">.</span>  <span style="font-style: italic"><span style="color: #9A1900"># assuming you have already made any changes that are system-specific</span></span>
<span style="color: #990000">.</span>/pfam_mpi<span style="color: #990000">.</span>sh   <span style="font-style: italic"><span style="color: #9A1900"># submits the job</span></span></tt></pre></div></div>
<div class="paragraph"><p>Wait until the run is complete:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt><span style="color: #009900">$TRANSDECODER_DIR</span>/util/ffindex_gather<span style="color: #990000">.</span>sh longest_orfs<span style="color: #990000">.</span>pep_out2    <span style="font-style: italic"><span style="color: #9A1900"># Tells you if any sequence has not been searched</span></span></tt></pre></div></div>
<div class="paragraph"><p>If something has not been processed (e.g. because walltime expired), resubmit the job; because of ffindex_gather, only sequences not already searched will be run.</p></div>
<div class="listingblock">
<div class="title">/pfam_mpi.sh</div>
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt><span style="color: #009900">$TRANSDECODER_DIR</span>/util/ffindex_gather<span style="color: #990000">.</span>sh longest_orfs<span style="color: #990000">.</span>pep_out2    <span style="font-style: italic"><span style="color: #9A1900"># check again.</span></span></tt></pre></div></div>
<div class="paragraph"><p>If all sequences have been searched, then an output file without the FFINDEX zero bytes will be produced:</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>TransDecoder -t target_transcripts<span style="color: #990000">.</span>fasta --search_pfam /path/to/Pfam-A<span style="color: #990000">.</span>hmm --MPI --workdir mytransdecoder_output --reuse --pfam_out mytransdecoder_output/longest_orfs<span style="color: #990000">.</span>pep_out2<span style="color: #990000">.</span>all<span style="color: #990000">.</span>db<span style="color: #990000">.</span>txt</tt></pre></div></div>
<h3 id="_pfam_output">PFAM output</h3><div style="clear:left"></div>
<div class="paragraph"><p>When TransDecoder is finished, you no longer require the mytransdecoder_output and can safely delete it. Before you do that, feel free to use the pfam_out file for <a href="http://trinotate.sf.net">Trinotate</a> to annotate your protein sequences. Also feel free to use these scripts as templates for other work: we (at CSIRO) routinely use this FFINDEX/MPI system to annotate the protein sequences with a variety of databases: a new software called <a href="http://jamps.sf.net">Just_Annotate_My_Proteins</a> (JAMp) is coming this (Southern) summer to a SourceForge site near you.</p></div>
</div>
<h2 id="_viewing_the_orf_predictions_in_a_genome_browser">Viewing the ORF predictions in a genome browser</h2>
<div class="sectionbody">
<div class="paragraph"><p><a href="http://genomeview.org">GenomeView</a> or <a href="http://www.broadinstitute.org/igv/">IGV</a> are  recommended for viewing the candidate ORFs in the context of the genome or the transcriptome.  Examples below show GenomeView in this context.</p></div>
<h3 id="_viewing_orfs_on_target_transcripts">Viewing ORFs on target transcripts</h3><div style="clear:left"></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>java -jar <span style="color: #009900">$GENOMEVIEW</span>/genomeview<span style="color: #990000">.</span>jar transcripts<span style="color: #990000">.</span>fasta transcripts<span style="color: #990000">.</span>fasta<span style="color: #990000">.</span>transdecoder<span style="color: #990000">.</span>bed</tt></pre></div></div>
<div class="paragraph"><p>If you lack a genome sequence and are working exclusively with the target transcripts, you can load the transcript fasta file and the ORF predictions (bed file) into GenomeView (see below).</p></div>
<div class="paragraph"><p><span class="image">
<img src="images/genomeview_trans.png" alt="Transcript shown with predicted ORF" />
</span></p></div>
<h3 id="_viewing_orfs_in_the_context_of_the_transcript_structures_on_the_genome">Viewing ORFs in the context of the transcript structures on the genome</h3><div style="clear:left"></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>java -jar <span style="color: #009900">$GENOMEVIEW</span>/genomeview<span style="color: #990000">.</span>jar <span style="font-weight: bold"><span style="color: #0000FF">test</span></span><span style="color: #990000">.</span>genome<span style="color: #990000">.</span>fasta transcripts<span style="color: #990000">.</span>bed transcripts<span style="color: #990000">.</span>fasta<span style="color: #990000">.</span>transdecoder<span style="color: #990000">.</span>genome<span style="color: #990000">.</span>bed</tt></pre></div></div>
<div class="paragraph"><p>The original cufflinks-based transcript structures are shown in black, and the predicted coding regions are shown in cyan.</p></div>
<div class="paragraph"><p><span class="image">
<img src="images/genomeview_cufflinks.png" alt="Cufflinks trans in GenomeView with predicted ORFs" />
</span></p></div>
</div>
<h2 id="_technical_support_and_project_announcements">Technical Support and Project Announcements</h2>
<div class="sectionbody">
<div class="paragraph"><p>Subscribe to our mailing list at <a href="https://lists.sourceforge.net/lists/listinfo/transdecoder-users">https://lists.sourceforge.net/lists/listinfo/transdecoder-users</a>.</p></div>
<div class="paragraph"><p>Email us at <a href="mailto:transdecoder-users@lists.sf.net">transdecoder-users@lists.sf.net</a>[mailto:transdecoder-users@lists.sf.net]</p></div>
</div>
</div>
<div id="footnotes"><hr /></div>
<div id="footer">
<div id="footer-text">
Last updated 2014-01-16 21:10:38 EST
</div>
</div>
</body>
</html>
